File: leet_BFS_1161_level_sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxLevelSum(root: TreeNode | null): number {
    if (!root)
        return 0
    let dq:TreeNode[] = [root]
    let record = Number.MIN_SAFE_INTEGER
    let level:number = 0
    let res:number = 0
    while (dq.length) {
        ++level
        let curr:number = 0
        let len:number = dq.length
        while (len--) {
            let node:TreeNode = dq.shift()
            curr += node.val
            if (node.left) {
                dq.push(node.left)
            }
            if (node.right) {
                dq.push(node.right)
            }
        }
        if (curr > record) {
            res = level
        }
        record = Math.max(curr, record)
    }
    return res
};

File: leet_BFS_1161_level_sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxLevelSum(root *TreeNode) int {
    if root == nil {
        return 0
    }
    D := []*TreeNode {root}
    record := math.MinInt32
    level := 0
    res := 0
    for len(D) > 0 {
        level++
        curr := 0
        size := len(D)
        i := -1
        for i < size - 1 {
            i++
            node := D[0]
            D = D[1:]
            curr += node.Val
            if node.Left != nil {
                D = append(D, node.Left)
            }
            if node.Right != nil {
                D = append(D, node.Right)
            }
        }
        if curr > record {
            res = level
        }
        record = int(math.Max(float64(curr), float64(record)))
        // record = math.Max(curr, res)
        // cannot use math.Max(curr, res) (value of type float64) as type int in assignment (solution.go)
    }
    return res
}
